### PR TITLE
Order resolved definitions from more to less specific

### PR DESCRIPTION
**Pull request by uarlouski**

### CODE REVIEW

File: vividus-studio-server/vividus-studio-plugin/src/main/java/org/vividus/studio/plugin/service/StepDefinitionResolver.java [hunk 1]
```diff
@@ -120,19 +120,43 @@ public void setStepDefinitions(Collection<StepDefinition> stepDefinitions)
         this.stepDefinitions.addAll(stepDefinitions);
     }
 
-    private Function<StepDefinition, Optional<ResolvedStepDefinition>> tryToResolve(Step step)
+    private Stream<ResolvedStepDefinition> resolve(Step step, List<StepDefinition> definitions, boolean expectAtMostOne)
```
Comment (valfirst) [inline vividus-studio-server/vividus-studio-plugin/src/main/java/org/vividus/studio/plugin/service/StepDefinitionResolver.java:123]:
```suggestion
    private Stream<ResolvedStepDefinition> resolve(Step step, List<StepDefinition> definitions, boolean limitResultsToOne)
```

File: vividus-studio-server/vividus-studio-plugin/src/main/java/org/vividus/studio/plugin/service/StepDefinitionResolver.java [hunk 2]
```diff
@@ -120,19 +120,44 @@ public void setStepDefinitions(Collection<StepDefinition> stepDefinitions)
...
         {
-            MatchOutcome outcome = TokenMatcher.match(step.getValue().strip(), def.getMatchTokens());
-            if (!outcome.isMatch())
+            List<String> matchTokens = definition.getMatchTokens();
+            MatchOutcome outcome = TokenMatcher.match(stepValue, matchTokens);
+
+            if (outcome.isMatch())
             {
-                return Optional.empty();
+                if (outcome.getTokenIndex() == matchTokens.size() - 1 && limitResultsToOne)
```
Comment (ikalinin1) [inline vividus-studio-server/vividus-studio-plugin/src/main/java/org/vividus/studio/plugin/service/StepDefinitionResolver.java:136]:
```suggestion
                if (limitResultsToOne && outcome.getTokenIndex() == matchTokens.size() - 1)
```

File: vividus-studio-server/vividus-studio-plugin/src/main/java/org/vividus/studio/plugin/service/StepDefinitionResolver.java [hunk 3]
```diff
@@ -120,19 +120,44 @@ public void setStepDefinitions(Collection<StepDefinition> stepDefinitions)
...
-            ResolvedStepDefinition resolved = new ResolvedStepDefinition(step.getLineIndex(), outcome.getTokenIndex(),
-                    outcome.getSubToken(), outcome.getArgIndices(), def);
-            return Optional.of(resolved);
-        };
+        }
+
+        Collections.sort(matchedDefinitions,
+                Comparator.comparing(e -> e.getValue().getTokenIndex(), Comparator.reverseOrder()));
+
+        if (!matchedDefinitions.isEmpty() && limitResultsToOne)
```
Comment (ikalinin1) [inline vividus-studio-server/vividus-studio-plugin/src/main/java/org/vividus/studio/plugin/service/StepDefinitionResolver.java:148]:
```suggestion
        if (limitResultsToOne && !matchedDefinitions.isEmpty())
```

File: vividus-studio-server/vividus-studio-plugin/src/main/java/org/vividus/studio/plugin/service/StepDefinitionResolver.java [hunk 4]
```diff
@@ -105,12 +109,8 @@ public Stream<ResolvedStepDefinition> resolve(String documentIdentifier)
...
 
         return steps.stream()
-                    .map(step -> groupedStepDefinitions.get().get(step.getType())
-                        .stream()
-                        .map(tryToResolve(step))
-                        .filter(Optional::isPresent)
-                        .findFirst())
-                    .flatMap(Optional::stream)
+                    .map(step -> resolve(step, groupedStepDefinitions.get().get(step.getType()), true)
+                            .findFirst())
```
Comment (ikalinin1) [inline vividus-studio-server/vividus-studio-plugin/src/main/java/org/vividus/studio/plugin/service/StepDefinitionResolver.java:113]:
```suggestion
                    .flatMap(step -> resolve(step, groupedStepDefinitions.get().get(step.getType()), true))
```

File: vividus-studio-server/vividus-studio-plugin/src/main/java/org/vividus/studio/plugin/service/StepDefinitionResolver.java [hunk 5]
```diff
@@ -72,13 +72,17 @@ public StepDefinitionResolver(TextDocumentProvider textDocumentProvider)
     public Stream<ResolvedStepDefinition> resolveAtPosition(String documentIdentifier, Position position)
     {
         return findStep(textDocumentProvider.getTextDocument(documentIdentifier), position)
-                .map(step -> groupedStepDefinitions.get().get(step.getType())
-                        .stream()
-                        .map(tryToResolve(step))
-                        .flatMap(Optional::stream))
+                .map(step -> resolve(step, groupedStepDefinitions.get().get(step.getType()), false))
```
Comment (ikalinin1) [inline vividus-studio-server/vividus-studio-plugin/src/main/java/org/vividus/studio/plugin/service/StepDefinitionResolver.java:75]:
```suggestion
                .flatMap(step -> resolve(step, groupedStepDefinitions.get().get(step.getType()), false))
```

Comment (uarlouski) [reply vividus-studio-server/vividus-studio-plugin/src/main/java/org/vividus/studio/plugin/service/StepDefinitionResolver.java:75]: the initial strem here is from optional element so it won't work

Comment (ikalinin1) [reply vividus-studio-server/vividus-studio-plugin/src/main/java/org/vividus/studio/plugin/service/StepDefinitionResolver.java:75]: java.util.Optional#stream
?

**Review by valfirst — APPROVED** [review]

**Review by ikalinin1 — APPROVED** [review]

**Pull request merged**